scipy/linalg/_decomp_update.pyx

```diff
@@ -0,0 +1,2416 @@
+        else:
+            raise ValueError('1 <= u.ndim <= 2')
+    else:
+        raise ValueError('q must be either F or C contig')
+
+cdef validate_array(cnp.ndarray a, bint chkfinite):
+    # here we check that a has positive strides and that its size is small
+    # enough to fit in into an int, as BLAS/LAPACK require
+    cdef bint copy = False
+    cdef bint too_large = False
+    cdef int j
+
+    for j in range(a.ndim):
+        if a.strides[j] <= 0 or \
+                (a.strides[j] / a.descr.itemsize) >= limits.INT_MAX:
+            copy = True
+        if a.shape[j] >= limits.INT_MAX:
+            raise ValueError('Input array to large for use with BLAS')
+
+    if chkfinite:
+        if not np.isfinite(a).all():
+            raise ValueError('array must not contain infs or NaNs')
+
+    if copy:
+            return PyArray_FromArraySafe(a, NULL, cnp.NPY_F_CONTIGUOUS)
+    return a
+
+cdef validate_qr(object q0, object r0, bint overwrite_q, int q_order,
+                 bint overwrite_r, int r_order, bint chkfinite):
+    cdef cnp.ndarray Q
+    cdef cnp.ndarray R
+    cdef int typecode
+    cdef bint economic = False
+
+    q_order |= cnp.NPY_BEHAVED_NS | cnp.NPY_ELEMENTSTRIDES  
+    r_order |= cnp.NPY_BEHAVED_NS | cnp.NPY_ELEMENTSTRIDES
+
+    if not overwrite_q:
+        q_order |= cnp.NPY_ENSURECOPY
+
+    if not overwrite_r:
+        r_order |= cnp.NPY_ENSURECOPY
+
+    # in the interests of giving better error messages take any number of
+    # dimensions here.
+    Q = PyArray_CheckFromAny(q0, NULL, 0, 0, q_order, NULL)
+    R = PyArray_CheckFromAny(r0, NULL, 0, 0, r_order, NULL)
+
+    if Q.ndim != 2 or R.ndim != 2:
+        raise ValueError('Q and R must be 2d')
+
+    typecode = cnp.PyArray_TYPE(Q)
+
+    if typecode != cnp.PyArray_TYPE(R):
+        raise ValueError('q and r must have the same type')
+
+    if not (typecode == cnp.NPY_FLOAT or typecode == cnp.NPY_DOUBLE 
+            or typecode == cnp.NPY_CFLOAT or typecode == cnp.NPY_CDOUBLE):
+        raise ValueError('only floatingcomplex arrays supported')
+
+    # we support MxM MxN and MxN NxN
+    if Q.shape[1] != R.shape[0]:
+        raise ValueError('Q and R do not have compatible shapes')
+
+    # so one or the other or both should be square.
+    if Q.shape[0] != Q.shape[1] and R.shape[0] == R.shape[1]:
+        economic = True
+    elif Q.shape[0] != Q.shape[1]:
+        raise ValueError('bad shapes.')
+
+    Q = validate_array(Q, chkfinite)
+    R = validate_array(R, chkfinite)
+
+    return Q, R, typecode, Q.shape[0], R.shape[1], economic
+ 
+cdef void* extract(cnp.ndarray a, int* as):
+    if a.ndim == 2:
+        as[0] = a.strides[0] / cnp.PyArray_ITEMSIZE(a)
+        as[1] = a.strides[1] / cnp.PyArray_ITEMSIZE(a)
+    elif a.ndim == 1:
+        as[0] = a.strides[0] / cnp.PyArray_ITEMSIZE(a)
+        as[1] = 0
+    return cnp.PyArray_DATA(a)
+
+@cython.embedsignature(True)
```

Comment:
cool, I wasn't aware this existed. we can use this in more places where we now maintain signatures manually